src/hmem_ipc_cache.c

```diff
@@ -0,0 +1,166 @@
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <ofi_mr.h>
+#include <ofi_util.h>
+#include <ofi_hmem.h>
+
+static int ipc_cache_add_region(struct ofi_mr_cache *cache, struct ofi_mr_entry *entry)
+{
+	int ret;
+	ret = ofi_hmem_open_handle(entry->info.iface, (void **)&entry->info.ipc_handle,
+				   entry->info.device, &entry->info.ipc_mapped_addr);
+	if (OFI_UNLIKELY(ret)) {
+		/*
+		 * There is a chance we can get the -FI_EALREADY from the
+		 * ofi_hmem_open_handle call. For cuda, the case that this
+		 * can happen is as follows. The sender gets handle from a
+		 * block of memory. Then the sending side frees the memory.
+		 * The sending side then cudaMalloc again and gets the same base
+		 * address. However, it cudaMalloc a block that is larger than
+		 * the one in the cache. The cache will return that memory is not
+		 * found and the ofi_hmem_open_handle will be called again.
+		 * However, that will fail with cudaErrorAlreadyMapped.
+		 * Therefore we need to unmap all overlapping regions and retry.
+		 * ofi_mr_cache_search already move all overlapping regions to
+		 * the dead_region_list via `until_mr_uncache_entry`.
+		 * We need to flush the cache to purge the entries
+		 * and close the handles.
+		 */
+		if (OFI_LIKELY(ret == -FI_EALREADY)) {
+			ofi_mr_cache_flush(cache, false);
+			ret = ofi_hmem_open_handle(entry->info.iface, (void **)&entry->info.ipc_handle,
+						   entry->info.device, &entry->info.ipc_mapped_addr);
+			if (OFI_UNLIKELY(ret)) {
+				FI_WARN(&core_prov, FI_LOG_CORE,
+					"Failed to open hmem handle, addr: %p, len: %lu\n",
+					entry->info.iov.iov_base, entry->info.iov.iov_len);
+			}
+		} else {
+			FI_WARN(&core_prov, FI_LOG_CORE,
+				"Failed to open hmem handle, addr: %p, len: %lu\n",
+				entry->info.iov.iov_base, entry->info.iov.iov_len);
+		}
+	}
```

Comment:
This function should be:
```
int ret;

ret = ofi_hmem_open...(...);
if (ret == -FI_EALREADY) {
    ofi_mr_flush...
    ret = ofi_hmem_open...
}
if (ret) {
    FI_WARN(...);
}
return ret;
```
Do not use OFI_UNLIKELY.  Programmers almost always get that wrong, and the errors here can happen.